### ISSUE TITLE
Fail on duplicate test names

### ISSUE THREAD

**Issue by Vampire** [bats-core/bats-core#278]
I fell into the trap that I accidentally had same-named tests.
This caused one of tests being run twice and the other not at all.
I consider this rather critical, as it can hide failures and you might wonder what goes on while all tests are green.

I've seen that there is already a detection for duplicate test names and a warning printed at the start of the run.
But if you start the tests, then wait until they are finished and then look at the result, this warning can easily be missed (I did so).
I'd like to suggest to change this to a hard error, as it hides potential test failures by not executing the code that should be executed.

**Comment by jasonkarns** [comment]
I can't think of a scenario where a user would intentionally use the same name, so I'm inclined to agree that we should hard-error.

**Comment by Vampire** [comment]
Actually I fell in that trap myself right away :-( svn-all-fast-export/svn2git#106.
As I have a wrapper around bats with some options anyway I added an output screening now that fails the test run in the end if there were duplicate names.

For me it is ok and maybe even better if there is no instant fail, as the tests that are run are valid nonetheless, just that one task is run twice and one isn't run at all.
So just like a failing test does not immediately abort, aborting after the tests are run would be desirable, as you could in the same run also see other failing tests.

But anything is better than only having a warning imho :-)

**Comment by sublimino** [comment]
I support having it fail fast, to avoid a safe sense of safety. Good point @Vampire :+1:

**Comment by jasonkarns** [comment]
@sublimino what do most other test runners do in this scenario? I'd prefer to behave similarly unless we have a strong reason not to.

**Comment by sublimino** [comment]
@jasonkarns from browsing a few other repos a lot of other test runners can't detect it as they define native functions, which are overwritten by the last definition (unless defining immutable variables).

I've been stung by this plenty of times so err on the side of obviousness (and trying to reclaim hours back!).

**Comment by jasonkarns** [comment]
I just quickly checked rspec and minitest (or, at least minitest within rails).

rspec happily duplicates the test, so both copies run as if they had different names. no issue.

minitest, (when using the `test "name"` DSL from ActiveSupport), exhibits the same restriction that bats has. That is, the second attempt to define a dupicate test name causes a collision. Minitest allows the entire suite to run, and treats the collision as an Error (which is different from a Failure). Errors are essentially mis-uses of the test framework, or exceptions that are thrown during the test that are not expected. They are distinct from Failures which are assertions that do not pass.

So, TL;DR:
- one framework _supports_ duplicate test names (uniq-ifying them somehow automatically)
- another framework does not allow duplicate test names, but does _NOT_ hard-error the test run. It completes, and just treats the duplicate test as errored/failed/broken.

We could probably audit some JS testing frameworks as well, but I'm mostly convinced myself that we should just report the test as failed instead of aborting the entire suite.
